horovod/common/ops/cuda_operations.cc

```diff
@@ -187,5 +187,101 @@ Status CUDAAllreduce::FinalizeCUDAQueue(const std::vector<TensorTableEntry>& ent
   return Status::InProgress();
 }
 
+CUDAAllgather::CUDAAllgather(CUDAContext* context,
+                             HorovodGlobalState* global_state)
+    : AllgatherOp(global_state), cuda_context_(context) {}
+
+bool CUDAAllgather::Enabled(const ParameterManager& param_manager,
+                            const std::vector<TensorTableEntry>& entries,
+                            const Response& response) const {
+  return entries[0].device != CPU_DEVICE_ID;
+}
+
+void CUDAAllgather::MemcpyEntryInFusionBuffer(const std::vector<TensorTableEntry>& entries,
+                                              const TensorTableEntry& e, void* buffer_data_at_offset) {
+  auto& first_entry = entries[0];
+  auto cuda_result = cudaMemcpyAsync(buffer_data_at_offset, e.tensor->data(),
+                                     (size_t) e.tensor->size(), cudaMemcpyDeviceToDevice,
+                                     cuda_context_->streams[global_state_->current_nccl_stream][first_entry.device]);
+  cuda_context_->ErrorCheck("cudaMemcpyAsync", cuda_result);
+}
+
+void CUDAAllgather::MemcpyEntryOutFusionBuffer(const std::vector<TensorTableEntry>& entries,
+                                               const void* buffer_data_at_offset, TensorTableEntry& e,
+                                               int64_t entry_offset, size_t entry_size) {
+  auto& first_entry = entries[0];
+  auto cuda_result = cudaMemcpyAsync((int8_t*)e.output->data() + entry_offset, buffer_data_at_offset,
+                                     entry_size, cudaMemcpyDeviceToDevice,
+                                     cuda_context_->streams[global_state_->current_nccl_stream][first_entry.device]);
+  cuda_context_->ErrorCheck("cudaMemcpyAsync", cuda_result);
+}
+
+void CUDAAllgather::InitCUDA(const std::vector<TensorTableEntry>& entries) {
```

Comment:
These three methods are copied directly from CUDAAllreduce.  I want to de-dupe but I'm looking for guidance on code style:

1.  Composition - shared class similar to CUDAContext that encapsulates functionality as member variable.
1.  Multiple inheritance - class like `CUDAOp` and then we have `class CUDAAllreduce : public AllreduceOp, public CUDAOp`
1.  Templated inheritance - class like `template <typename Base> class CUDAOp : public Base` and then we have `class CUDAAllreduce : public CUDAOp<AllreduceOp>`
1.  Leave the code duplicated?

Reply:
Good question!  Looking through the code and how it's used, my suggestion would be to go with composition.  We could roll the three methods into a new class called `CUDAOpContext`.  Then, instead of subclasses like NCCLAllreduce and DDLAllreduce calling `InitCUDAQueue`, we have them call those methods on a protected member `cuda_op_context_`.  What do you think?

Reply:
Yes, I like this approach.  I'll add that and update the pull request!